ls1intum/Artemis
# Plagiarism Checks with Progress Feedback

### Checklist

### Motivation and Context


The progress of the "check plagiarism" action may be lengthy. Currently, the user only sees a spinner.

### Description

This PR extends that and adds some progress messages to notify the current state of the plagiarism detection progress so that the user can get more information. It also removes a bit of duplicated code.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create exercises of types programming, modeling, and text
4. Participate in them was different users and try to submit the same things to trigger the plagiarism
5. Navigate to the exercise detail page and click on "check plagiarism"
6. Depending on the amount of participation, you should be able to receive progress messages.

### Test Coverage



ModelingPlagiarismDetectionService.java 94%
PlagiarismWebsocketService.java 85%
ProgrammingPlagiarismDetectionService.java 80%
TextPlagiarismDetectionService.java 92%
ProgrammingExerciseExportService.java 67%
ProgrammingExerciseResource.java 91%
plagiarism-inspector.component.ts 87.93%
### Screenshots



Programming exercises:

![1](https://user-images.githubusercontent.com/11298674/116678976-43164700-a9aa-11eb-9a16-bf7bcb3c42ba.gif)

Text Exercises:

![2](https://user-images.githubusercontent.com/11298674/116678992-48739180-a9aa-11eb-8ac7-cd3e4edcecee.gif)
